libcudacxx/include/cuda/std/detail/libcxx/include/__utility/pair.h

```diff
@@ -69,707 +70,704 @@
+    tuple<_Args1...>& __first_args,
+    tuple<_Args2...>& __second_args,
+    __tuple_indices<_I1...>,
+    __tuple_indices<_I2...>);
+};
 
-    _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX17
-    pair& operator=(__conditional_t<
-                        is_move_assignable<first_type>::value &&
-                        is_move_assignable<second_type>::value,
-                    pair, __nat>&& __p)
-        noexcept(is_nothrow_move_assignable<first_type>::value &&
-                   is_nothrow_move_assignable<second_type>::value)
-    {
-        first = _CUDA_VSTD::forward<first_type>(__p.first);
-        second = _CUDA_VSTD::forward<second_type>(__p.second);
-        return *this;
-    }
+// We need to ensure that a reference type, which would inhibit the implicit copy assignment still works
+template <class _T1, class _T2>
+struct __pair_base<_T1, _T2, true>
+{
+  _T1 first;
+  _T2 second;
+
+  template <class _Constraints                                                 = __pair_constraints<_T1, _T2>,
+            __enable_if_t<_Constraints::__implicit_default_constructible, int> = 0>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr __pair_base() noexcept(
+    _LIBCUDACXX_TRAIT(is_nothrow_default_constructible, _T1)
+    && _LIBCUDACXX_TRAIT(is_nothrow_default_constructible, _T2))
+      : first()
+      , second()
+  {}
+
+  template <class _Constraints                                                 = __pair_constraints<_T1, _T2>,
+            __enable_if_t<_Constraints::__explicit_default_constructible, int> = 0>
+  _LIBCUDACXX_INLINE_VISIBILITY explicit constexpr __pair_base() noexcept(
+    _LIBCUDACXX_TRAIT(is_nothrow_default_constructible, _T1)
+    && _LIBCUDACXX_TRAIT(is_nothrow_default_constructible, _T2))
+      : first()
+      , second()
+  {}
+
+  constexpr __pair_base(const __pair_base&) = default;
+  constexpr __pair_base(__pair_base&&)      = default;
+
+  _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __pair_base& operator=(
+    __conditional_t<_LIBCUDACXX_TRAIT(is_copy_assignable, _T1) && _LIBCUDACXX_TRAIT(is_copy_assignable, _T2),
+                    __pair_base,
+                    __nat> const& __p) noexcept(_LIBCUDACXX_TRAIT(is_nothrow_move_assignable, _T1)
+                                                && _LIBCUDACXX_TRAIT(is_nothrow_move_assignable, _T2))
+  {
+    first  = __p.first;
+    second = __p.second;
+    return *this;
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 __pair_base& operator=(
+    __conditional_t<_LIBCUDACXX_TRAIT(is_move_assignable, _T1) && _LIBCUDACXX_TRAIT(is_move_assignable, _T2),
+                    __pair_base,
+                    __nat>&& __p) noexcept(_LIBCUDACXX_TRAIT(is_nothrow_move_assignable, _T1)
+                                           && _LIBCUDACXX_TRAIT(is_nothrow_move_assignable, _T2))
+  {
+    first  = _CUDA_VSTD::forward<_T1>(__p.first);
+    second = _CUDA_VSTD::forward<_T2>(__p.second);
```

Comment:
This warrants a comment, because at first I couldn't figure out what's going on here, until I exclaimed "oh, references!".